### PR TITLE
main,bind: Extended gopy:name control, auto PEP8 naming, and property docstring support

### PR DESCRIPTION
**Pull request by justinfx**
This merge improves the support for controlling the generates names of functions, methods, and properties in the python code. The "gopy:name" tag support has been extended to struct fields and updated to also support single line comments:

```go
type MyStruct struct {
    // gopy_name: my_field
    // This is my field.
    MyField int

    // This is my second field
   // gopy_name: my_field2
   MyField2 int
}
```

A new optional feature flat `-rename` has been added to apply automatic PEP8 snake_case name transformations to functions, methods, properties, and var getter/setters.  This makes it more convenient than having to annotate every symbol in Go source with a `gopy:name` tag if we only want CamelCase converted to snake_case to be more "pythonic".

And lastly, as part of parsing the comments for struct fields, the docstring is actually also applied to the python property getter in the generated code.

Objectives:

1. Is the `rename` flag and auto naming feature useful and complete enough?
2. Is the approach to parsing the struct field comments correct? I tried to review the existing way that method comments are propagated but it seemed fields were a bit different.

### CODE REVIEW

File: _examples/rename/rename.go [hunk 1]
```diff
@@ -5,15 +5,38 @@
 // package rename tests changing the names of methods and functions
 package rename
 
-// gopy:name say_hi
+// gopy:name say_hi_fn
+// Comment follows the tag and function shoudl be renamed
```
Comment (sbinet) [inline _examples/rename/rename.go:9]:
```suggestion
// Comment follows the tag and function should be renamed
```

File: _examples/rename/rename.go [hunk 2]
```diff
@@ -5,15 +5,38 @@
...
+// to say_hi_fn
 func SayHi() string {
 	return "hi"
 }
 
+// I should be renamed tp auto_renamed_func, when generated
```
Comment (sbinet) [inline _examples/rename/rename.go:15]:
```suggestion
// I should be renamed to auto_renamed_func, when generated
```

**Review by sbinet — COMMENTED** [review]
for struct fields, there's perhaps an argument about using struct-tags.
_e.g.:_
```go
type T struct {
    MyField int32 `gopy:"my_field"`
    Hidden  int32 `gopy:"_hidden"`
}
```

the advantage of using a `//gopy:name` directive might be on the godoc side (I suspect that kind of comment may not be displayed so these kind of internal details aren't leaked to the world, contrary to the struct-tag approach).

@rcoreilly ? any input?

**Comment by justinfx** [comment]
@sbinet in hindsight I should have used struct tags. It makes more sense and probably would have saved me the trouble of trying to locate the comments for the field. I was only thinking of making them the same as the other types, but a struct tag is better suited for settings like this that would be removed from the doc string.
Let me work up another commit to try and change this.

**Comment by rcoreilly** [comment]
Yeah I agree that on balance the field tag makes more sense.  All kinds of impl stuff is leaked through those so I don't think this is any different, and is idiomatic in my experience.

**Comment by justinfx** [comment]
I've now added struct tag parsing of `gopy:"new_name"`. Please let me know if you want me to remove the doc comment parsing logic entirely, or leave it in? I left it just for consistency with the other types, but the tag wins.

**Comment by rcoreilly** [comment]
the Go style is to have only one way to do things where possible, so probably remove it..

**Comment by justinfx** [comment]
I've removed the property getter/setter doc parsing for a name override. But I have left the struct field `getDoc` updates because they are needed to pull the comments to then set the property docstring in python

File: bind/utils.go [hunk 3]
```diff
@@ -192,19 +193,82 @@ func getGoVersion(version string) (int64, int64, error) {
...
+// a new python name. If the tag is not defined then the original
+// name is returned.
+// If the tag name is specified but is an invalid python identifier,
+// then an error is returned.
+func extractPythonNameFieldTag(gname, tag string) (string, error) {
+	const tagKey = "gopy"
+	if tag == "" {
+		return gname, nil
+	}
+	tags, err := structtag.Parse(tag)
```
Comment (sbinet) [inline bind/utils.go:239]: bringing this whole new dependency seems a bit overkill for something that could -I think- be achieved with just:
```go
v, ok := reflect.StructTag(tag).Lookup(tagKey)
```
couldn't it?

Comment (justinfx) [reply bind/utils.go]: Oops. I should have paid closer attention to what was already available in the `reflect` package. Fixed!

Comment (sbinet) [reply bind/utils.go]: no worries :)

**Review by sbinet — APPROVED** [review]
LGTM.

(I'll wait a bit for @rcoreilly before merging this one in)

**Comment by rcoreilly** [comment]
Yep looks good to me too -- thanks for the contributions!

**Pull request merged**